File: compiler/bench/ack.c
// Ackermann benchmark: ack(3,10)=8189, ack(4,1)=65533
#include <stdio.h>
long ack(long m, long n) {
    if (m == 0) return n + 1;
    if (n == 0) return ack(m - 1, 1);
    return ack(m - 1, ack(m, n - 1));
}
int main() {
    printf("%ld\n", ack(3, 10));
    printf("%ld\n", ack(4, 1));
    return 0;
}

File: compiler/bench/ack.fs
\ Ackermann function benchmark
\ Tests: ack(3,10)=8189, ack(4,1)=65533

: ack ( m n -- r ) recursive
  over 0= if nip 1+ exit then
  dup 0= if drop 1- 1 recurse exit then
  over 1- -rot 1- recurse recurse ;

: main
  3 10 ack . cr
  4 1 ack . cr
;
